If testng fails to create an instance, it should say which class fails

Here's the stacktrace.

The offending class was `org.ops4j.pax.exam.testng.listener.PaxExam` (testng was trying to instantiate `PaxExam`, however, it failed).

I suggest that failed reflection calls should add the relevant pointers to the class and suite names.

```
org.gradle.api.internal.tasks.testing.TestSuiteExecutionException: Could not complete execution for Gradle Test Executor 8.
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:63)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:135)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.NoClassDefFoundError: javax/transaction/NotSupportedException
	at java.base/java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredConstructors(Class.java:3296)
	at java.base/java.lang.Class.getConstructors(Class.java:2036)
	at org.testng.internal.objects.InstanceCreator.newInstance(InstanceCreator.java:47)
	at org.testng.ITestObjectFactory.newInstance(ITestObjectFactory.java:13)
	at org.testng.internal.objects.SimpleObjectDispenser.dispense(SimpleObjectDispenser.java:57)
	at org.testng.internal.objects.GuiceBasedObjectDispenser.dispense(GuiceBasedObjectDispenser.java:32)
	at org.testng.internal.DefaultListenerFactory.createListener(DefaultListenerFactory.java:29)
	at org.testng.TestRunner.initListeners(TestRunner.java:379)
	at org.testng.TestRunner.init(TestRunner.java:336)
	at org.testng.TestRunner.init(TestRunner.java:286)
	at org.testng.TestRunner.<init>(TestRunner.java:176)
	at org.testng.SuiteRunner$DefaultTestRunnerFactory.newTestRunner(SuiteRunner.java:635)
	at org.testng.SuiteRunner.init(SuiteRunner.java:217)
	at org.testng.SuiteRunner.<init>(SuiteRunner.java:114)
	at org.testng.TestNG.createSuiteRunner(TestNG.java:1308)
	at org.testng.TestNG.createSuiteRunners(TestNG.java:1284)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1133)
	at org.testng.TestNG.runSuites(TestNG.java:1071)
	at org.testng.TestNG.run(TestNG.java:1039)
	at org.gradle.api.internal.tasks.testing.testng.TestNGTestClassProcessor.runTests(TestNGTestClassProcessor.java:141)
	at org.gradle.api.internal.tasks.testing.testng.TestNGTestClassProcessor.stop(TestNGTestClassProcessor.java:90)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	... 25 more
Caused by: java.lang.ClassNotFoundException: javax.transaction.NotSupportedException
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:606)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:168)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 48 more
```

--- a/testng-core-api/src/main/java/org/testng/internal/objects/InstanceCreator.java
+++ b/testng-core-api/src/main/java/org/testng/internal/objects/InstanceCreator.java
@@ -1,5 +1,8 @@
 package org.testng.internal.objects;
 
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
 import java.util.Objects;
@@ -23,15 +26,13 @@ public final class InstanceCreator {
 
   public static <T> T newInstance(Class<T> clazz) {
     try {
-      return clazz.getDeclaredConstructor().newInstance();
-    } catch (IllegalAccessException
-        | InstantiationException
-        | ExceptionInInitializerError
-        | NoSuchMethodException
-        | SecurityException e) {
+      MethodHandle constructor =
+          MethodHandles.lookup().findConstructor(clazz, MethodType.methodType(void.class));
+      return (T) constructor.invoke();
+    } catch (IllegalAccessException | NoSuchMethodException | SecurityException e) {
       throw new TestNGException(CANNOT_INSTANTIATE_CLASS + clazz.getName(), e);
-    } catch (InvocationTargetException e) {
-      throw new TestNGException(CANNOT_INSTANTIATE_CLASS + clazz.getName(), e.getCause());
+    } catch (Throwable e) {
+      throw new TestNGException(CANNOT_INSTANTIATE_CLASS + clazz.getName(), e);
     }
   }
 
@@ -45,6 +46,9 @@ public final class InstanceCreator {
   }
 
   public static <T> T newInstance(Class<T> cls, Object... parameters) {
+    if (parameters.length == 0) {
+      return newInstance(cls);
+    }
     Constructor<T> ctor = null;
     for (Constructor<?> c : cls.getConstructors()) {
       // Just comparing parameter array sizes. Comparing the parameter types

--- a/testng-core/src/test/java/test/listeners/ListenersTest.java
+++ b/testng-core/src/test/java/test/listeners/ListenersTest.java
@@ -1,6 +1,8 @@
 package test.listeners;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 import java.io.IOException;
 import java.nio.file.Files;
@@ -23,6 +25,8 @@ import test.SimpleBaseTest;
 import test.listeners.issue2381.FactoryTestClassSample;
 import test.listeners.issue2381.SampleGlobalListener;
 import test.listeners.issue2381.SampleTransformer;
+import test.listeners.issue2578.ListenerWithMissingConstructorDependency;
+import test.listeners.issue2578.MissingDependencyClassLoader;
 import test.listeners.issue2638.DummyInvokedMethodListener;
 import test.listeners.issue2638.TestClassASample;
 import test.listeners.issue2638.TestClassBSample;
@@ -683,6 +687,20 @@ public class ListenersTest extends SimpleBaseTest {
     TestNG testng = create(ChildClassSample.class);
     testng.run();
     assertThat(testng.getStatus()).isZero();
+  }
+
+  @Test(description = "GITHUB-2578")
+  public void ensureListenerWithUnavailableConstructorDependencyDoesNotFail()
+      throws ClassNotFoundException {
+    MissingDependencyClassLoader classLoader = new MissingDependencyClassLoader();
+    Class<?> listenerClass = classLoader.loadClass(ListenerWithMissingConstructorDependency.class);
+    Class<?> testClass = classLoader.loadClass(test.listeners.issue2578.TestClassSample.class);
+    TestNG testng = create(testClass);
+
+    assertThatThrownBy(listenerClass::getConstructors)
+        .isInstanceOf(NoClassDefFoundError.class)
+        .hasCauseInstanceOf(ClassNotFoundException.class);
+    assertThatCode(testng::run).doesNotThrowAnyException();
   }
 
   private void setupTest(boolean addExplicitListener) {

--- a/testng-core/src/test/java/test/listeners/issue2578/ListenerWithMissingConstructorDependency.java
+++ b/testng-core/src/test/java/test/listeners/issue2578/ListenerWithMissingConstructorDependency.java
@@ -1,0 +1,10 @@
+package test.listeners.issue2578;
+
+import org.testng.ITestListener;
+
+public class ListenerWithMissingConstructorDependency implements ITestListener {
+
+  public ListenerWithMissingConstructorDependency() {}
+
+  public ListenerWithMissingConstructorDependency(MissingConstructorDependency dependency) {}
+}

--- a/testng-core/src/test/java/test/listeners/issue2578/MissingConstructorDependency.java
+++ b/testng-core/src/test/java/test/listeners/issue2578/MissingConstructorDependency.java
@@ -1,0 +1,3 @@
+package test.listeners.issue2578;
+
+public class MissingConstructorDependency {}

--- a/testng-core/src/test/java/test/listeners/issue2578/MissingDependencyClassLoader.java
+++ b/testng-core/src/test/java/test/listeners/issue2578/MissingDependencyClassLoader.java
@@ -1,0 +1,50 @@
+package test.listeners.issue2578;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+public class MissingDependencyClassLoader extends ClassLoader {
+
+  private static final String SAMPLE_PACKAGE = "test.listeners.issue2578.";
+  private static final String MISSING_DEPENDENCY = SAMPLE_PACKAGE + "MissingConstructorDependency";
+
+  public MissingDependencyClassLoader() {
+    super(MissingDependencyClassLoader.class.getClassLoader());
+  }
+
+  public Class<?> loadClass(Class<?> clazz) throws ClassNotFoundException {
+    return loadClass(clazz.getName());
+  }
+
+  @Override
+  protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+    if (MISSING_DEPENDENCY.equals(name)) {
+      throw new ClassNotFoundException(name);
+    }
+    if (name.startsWith(SAMPLE_PACKAGE)) {
+      Class<?> loadedClass = findLoadedClass(name);
+      if (loadedClass == null) {
+        loadedClass = findClass(name);
+      }
+      if (resolve) {
+        resolveClass(loadedClass);
+      }
+      return loadedClass;
+    }
+    return super.loadClass(name, resolve);
+  }
+
+  @Override
+  protected Class<?> findClass(String name) throws ClassNotFoundException {
+    String resourceName = name.replace('.', '/') + ".class";
+    try (InputStream inputStream = getParent().getResourceAsStream(resourceName)) {
+      if (inputStream == null) {
+        throw new ClassNotFoundException(name);
+      }
+      byte[] bytes = inputStream.readAllBytes();
+      return defineClass(name, bytes, 0, bytes.length);
+    } catch (IOException e) {
+      throw new ClassNotFoundException(name, e);
+    }
+  }
+}

--- a/testng-core/src/test/java/test/listeners/issue2578/TestClassSample.java
+++ b/testng-core/src/test/java/test/listeners/issue2578/TestClassSample.java
@@ -1,0 +1,11 @@
+package test.listeners.issue2578;
+
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+
+@Listeners(ListenerWithMissingConstructorDependency.class)
+public class TestClassSample {
+
+  @Test
+  public void testMethod() {}
+}

